libraries/rush-lib/src/logic/operations/OperationLifecycle.ts

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { AsyncSeriesWaterfallHook } from 'tapable';
+
+import type { ITerminal } from '@rushstack/node-core-library';
+import type { IOperationRunnerContext } from './IOperationRunner';
+import type { OperationStatus } from './OperationStatus';
+import type { IProjectDeps, ShellOperationRunner } from './ShellOperationRunner';
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import type { ProjectChangeAnalyzer } from '../ProjectChangeAnalyzer';
+
+export interface IOperationRunnerBeforeExecuteContext {
+  context: IOperationRunnerContext;
+  runner: ShellOperationRunner;
```

Comment:
Is it possible to achieve the functionalities of "uses a pool of long-running worker processes and communicates via IPC" by tapping the lifecycle? If so, that will be great to totally eliminate the inheritance for "OperationRunner" concept.

For example, "NullOperationRunner" can be consolidated to "ShellOperationRunner" with a "NullableCommandPlugin", while in the NullableCommandPlugin, tapping the "beforeExecuteAsync" lifecycle and early return when shellCommand is empty.

Reply:
There's a mechanics problem in that if you use `beforeExecute` to replace the implementation by returning a result, then you lose the invocation of `afterExecute`. We'd have to define an `execute` hook, at which point that's exactly the same as having `Operation` have the hooks directly and `IOperationRunner` becomes superfluous.

I'm actually the one who renamed `OperationRunner` to `ShellOperationRunner` and introduced the `IOperationRunner` interface in the first place, expressly to support the general scenario of there being operations for which "execute" is something other than "run this shell command". I took heavy advantage of this feature in a prototype plugin that performs the installation of npm dependencies on-demand as part of the build graph, for which some of the runners are "make this directory", "download this file to a buffer", "make these symlinks",  "postmessage a buffer to a worker thread to extract", etc.

As another example, I have had an interest in having "write to the build cache" be its own operation in the graph (with no consumers) so that it doesn't block downstream operations (on Windows in particular having to wait for the tarball can be quite slow). This particular scenario could be addressed by having a queue of outstanding build cache work in the CacheableOperationPlugin and waiting for it to finish in `afterExecuteOperations`, which I'll probably look into after this PR merges, since your refactor has made doing so a lot simpler.